Router v2 integration

saledjenic opened the pull request:
What's done in this PR:
- `GetSuggestedRoutesV2Async` exposed via `wallet` API which calculates the best route in an async way and notifies the client via `"wallet.suggested.routes"` signal
- Old wallet event with signal type `"wallet"` and event type "sign-transaction", like this `{"type":"wallet","event":{"type":"sing-transactions", "transactions":["0x123", ...] .... } }` is now  set to signal type  `"wallet.sign.transactions"` and looks like this `{"type":"wallet.sign.transactions","event":["0x123", ...] }`
- A ew event type added `"wallet.suggested.routes"`, it carries the best path (list of paths v2) or error (code and description)

saledjenic committed:
chore!: wallet related signals alignment

A signal type `wallet` which had an event `type` set to `sing-transactions` is now a new signal
type `wallt.sing.transactions` with event which represents list of transactions that need to be signed.

saledjenic committed:
feat_: new endpoint added for an async route/s calculation

- `GetSuggestedRoutesV2Async` calculates the route/s based on input parameters and sends
`wallet.suggested.routes` signal to notify a client.

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (43)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 388b4eefb251e326b205149633ca541ad7077d7c | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5426/1/) | 2024-06-26 07:52:06 | ~4 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240626-074753-388b4eefb251e326b205149633ca541ad7077d7c-pr5426.zip) |
| :heavy_check_mark: | 388b4eefb251e326b205149633ca541ad7077d7c | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5426/1/) | 2024-06-26 07:52:36 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240626-074753-388b4eefb251e326b205149633ca541ad7077d7c-pr5426.zip) |
| :heavy_check_mark: | 388b4eefb251e326b205149633ca541ad7077d7c | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5426/1/) | 2024-06-26 07:53:33 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240626-074753-388b4eefb251e326b205149633ca541ad7077d7c-pr5426.aar) |
| :heavy_check_mark: | 388b4eefb251e326b205149633ca541ad7077d7c | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5426/2/) | 2024-06-26 10:06:49 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240626-100434-388b4eefb251e326b205149633ca541ad7077d7c-pr5426.zip) |
| :heavy_check_mark: | 388b4eefb251e326b205149633ca541ad7077d7c | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5426/2/) | 2024-06-26 10:07:44 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240626-100434-388b4eefb251e326b205149633ca541ad7077d7c-pr5426.zip) |
| :heavy_check_mark: | 388b4eefb251e326b205149633ca541ad7077d7c | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5426/2/) | 2024-06-26 10:11:43 | ~7 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240626-100434-388b4eefb251e326b205149633ca541ad7077d7c-pr5426.aar) |
| | | | | | | |
| :heavy_check_mark: | 388b4eef | [#1](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5426/1/) | 2024-06-26 08:31:54 | ~43 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5426/1//consoleText) |
| :heavy_check_mark: | 388b4eef | [#2](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5426/2/) | 2024-06-26 10:48:56 | ~44 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5426/2//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 342e8470b85753895423c47630ce14516ece8081 | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5426/3/) | 2024-06-26 10:11:44 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240626-100813-342e8470b85753895423c47630ce14516ece8081-pr5426.zip) |
| :heavy_check_mark: | 342e8470b85753895423c47630ce14516ece8081 | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5426/3/) | 2024-06-26 10:12:06 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240626-100807-342e8470b85753895423c47630ce14516ece8081-pr5426.zip) |
| :heavy_check_mark: | 342e8470b85753895423c47630ce14516ece8081 | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5426/3/) | 2024-06-26 10:21:10 | ~9 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240626-101159-342e8470b85753895423c47630ce14516ece8081-pr5426.aar) |
| | | | | | | |
| :heavy_multiplication_x: | 342e8470 | [#3](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5426/3/) | 2024-06-26 11:29:53 | ~40 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5426/3//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 81601a306a2a122b389215729b80365444fd3a7e | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5426/4/) | 2024-06-26 13:34:29 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240626-133201-81601a306a2a122b389215729b80365444fd3a7e-pr5426.zip) |
| :heavy_check_mark: | 81601a306a2a122b389215729b80365444fd3a7e | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5426/4/) | 2024-06-26 13:36:04 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240626-133201-81601a306a2a122b389215729b80365444fd3a7e-pr5426.zip) |
| :heavy_check_mark: | 81601a306a2a122b389215729b80365444fd3a7e | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5426/4/) | 2024-06-26 13:37:31 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240626-133201-81601a306a2a122b389215729b80365444fd3a7e-pr5426.aar) |
| | | | | | | |
| :heavy_check_mark: | 81601a30 | [#4](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5426/4/) | 2024-06-26 14:17:15 | ~45 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5426/4//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 80266209 | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5426/5/) | 2024-06-26 19:51:27 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240626-194856-802662-pr5426.aar) |
| :heavy_check_mark: | 80266209 | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5426/5/) | 2024-06-26 19:52:13 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240626-194856-802662-pr5426.zip) |
| :heavy_check_mark: | 80266209 | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5426/5/) | 2024-06-26 19:53:05 | ~4 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240626-194859-802662-pr5426.zip) |
| :heavy_check_mark: | 80266209 | [#5](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5426/5/) | 2024-06-26 20:31:29 | ~42 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5426/5//consoleText) |
| | | | | | | |
| :heavy_check_mark: | d3626039 | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5426/6/) | 2024-06-27 09:41:55 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240627-093934-d36260-pr5426.aar) |
| :heavy_check_mark: | d3626039 | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5426/6/) | 2024-06-27 09:43:48 | ~4 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240627-093938-d36260-pr5426.zip) |
| :heavy_multiplication_x: | d3626039 | [#6](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5426/6/) | 2024-06-27 09:45:43 | ~6 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5426/6//consoleText) |
| | | | | | | |
| :heavy_check_mark: | ac23170e | [#7](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5426/7/) | 2024-06-27 11:06:42 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240627-110439-ac2317-pr5426.aar) |
| :heavy_check_mark: | ac23170e | [#7](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5426/7/) | 2024-06-27 11:06:56 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240627-110439-ac2317-pr5426.zip) |
| :heavy_multiplication_x: | ac23170e | [#7](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5426/7/) | 2024-06-27 11:38:49 | ~34 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5426/7//consoleText) |
| :heavy_check_mark: | ac23170e | [#8](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5426/8/) | 2024-06-27 12:30:36 | ~41 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5426/8//consoleText) |
| :x: | ac23170e | [#12](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5426/12/) | 2024-06-27 14:09:39 | ~7 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5426/12/consoleText) |
| | | | | | | |
| :heavy_check_mark: | bfd16186 | [#8](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5426/8/) | 2024-06-27 15:04:39 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240627-150230-bfd161-pr5426.aar) |
| :heavy_check_mark: | bfd16186 | [#8](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5426/8/) | 2024-06-27 15:04:57 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240627-150234-bfd161-pr5426.zip) |
| :heavy_check_mark: | bfd16186 | [#9](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5426/9/) | 2024-06-27 15:44:16 | ~41 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5426/9//consoleText) |
| | | | | | | |
| :heavy_check_mark: | d047f0f5 | [#9](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5426/9/) | 2024-06-28 06:28:33 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240628-062623-d047f0-pr5426.aar) |
| :heavy_check_mark: | d047f0f5 | [#9](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5426/9/) | 2024-06-28 06:28:46 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240628-062627-d047f0-pr5426.zip) |
| :heavy_check_mark: | d047f0f5 | [#15](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5426/15/) | 2024-06-28 06:30:31 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240628-062627-d047f0-pr5426.zip) |
| :heavy_check_mark: | d047f0f5 | [#10](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5426/10/) | 2024-06-28 07:08:27 | ~41 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5426/10//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 7398100b | [#10](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5426/10/) | 2024-06-28 13:55:06 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240628-135246-739810-pr5426.aar) |
| :heavy_check_mark: | 7398100b | [#10](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5426/10/) | 2024-06-28 13:55:14 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240628-135250-739810-pr5426.zip) |
| :heavy_check_mark: | 7398100b | [#16](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5426/16/) | 2024-06-28 13:55:45 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240628-135246-739810-pr5426.zip) |
| :heavy_multiplication_x: | 7398100b | [#11](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5426/11/) | 2024-06-28 13:59:31 | ~6 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5426/11//consoleText) |
| | | | | | | |
| :heavy_check_mark: | b01af610 | [#11](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5426/11/) | 2024-06-30 20:49:36 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240630-204707-b01af6-pr5426.aar) |
| :heavy_check_mark: | b01af610 | [#11](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5426/11/) | 2024-06-30 20:49:39 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240630-204712-b01af6-pr5426.zip) |
| :heavy_check_mark: | b01af610 | [#17](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5426/17/) | 2024-06-30 20:50:23 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240630-204712-b01af6-pr5426.zip) |
| :heavy_check_mark: | b01af610 | [#12](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5426/12/) | 2024-06-30 21:31:15 | ~44 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5426/12//consoleText) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 654ba937 | [#18](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5426/18/) | 2024-07-01 11:24:37 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240701-112014-654ba9-pr5426.zip) |
| :heavy_check_mark: | 654ba937 | [#12](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5426/12/) | 2024-07-01 11:32:10 | ~11 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240701-112014-654ba9-pr5426.zip) |
| :heavy_check_mark: | 654ba937 | [#12](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5426/12/) | 2024-07-01 11:35:37 | ~15 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240701-112014-654ba9-pr5426.aar) |
| :heavy_check_mark: | 654ba937 | [#13](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5426/13/) | 2024-07-01 12:12:07 | ~51 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5426/13//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 9759e74d | [#19](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5426/19/) | 2024-07-01 11:49:09 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240701-114608-9759e7-pr5426.zip) |
| :heavy_check_mark: | 9759e74d | [#13](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5426/13/) | 2024-07-01 11:52:14 | ~6 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240701-114608-9759e7-pr5426.zip) |
| :heavy_check_mark: | 9759e74d | [#13](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5426/13/) | 2024-07-01 11:53:22 | ~7 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240701-114604-9759e7-pr5426.aar) |
| :heavy_check_mark: | 9759e74d | [#14](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5426/14/) | 2024-07-01 12:57:02 | ~44 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5426/14//consoleText) |

saledjenic commented:
@ilmotta changes done here should not affect the mobile app since it doesn't respond to `sing-transactions` signal (you will start using it when you implement keycard tx signing).

dlipicar commented on services/wallet/router/router_v2.go:
```diff
@@ -417,7 +418,68 @@ func validateInputData(input *RouteInputParams) error {
+			signal.SendWalletEvent(signal.SuggestedRoutes, errors.CreateErrorResponseFromError(err))
+		}
+
+		if r.isBusy() {
```
Hmm if multiple concurrent requests are not supported, we most likely want to cancel the old request instead of returning an error for the new one, right? The typical scenario is that the user changes some parameter when setting up a Send/Bridge/Swap and the client performs a new `GetSuggestedRoutes` call before the results of the previous one arrive.
The [scheduler](https://github.com/status-im/status-go/blob/develop/services/wallet/async/scheduler.go) should help with that, it's what we use for `[FilterActivity](https://github.com/status-im/status-go/blob/49eaabaca5100368c5b39fb8c107aad2535371e5/services/wallet/activity/service.go#L123C40-L123C54)` or [some collectibles endpoints](https://github.com/status-im/status-go/blob/49eaabaca5100368c5b39fb8c107aad2535371e5/services/wallet/collectibles/service.go#L301). In those cases we support multiple concurrent calls from different sources, but here you could use `scheduler` instead of `multi_client_scheduler`.

saledjenic replied:
Yes, makes sense, that's what I saw when I finished the implementation. Updating, thanks.

dlipicar commented on services/wallet/router/errors.go:
```diff
@@ -6,6 +6,7 @@ import (
 
 // Abbreviation `WR` for the error code stands for Wallet Router
 var (
+	ErrRouterIsBusy                              = &errors.ErrorResponse{Code: errors.ErrorCode("WR-000"), Details: "the router is busy"}
```
unused after the changes to use the scheduler, right?

dlipicar commented on services/wallet/router/router_v2.go:
```diff
@@ -417,6 +418,17 @@ func validateInputData(input *RouteInputParams) error {
+	}, func(ctx context.Context) (interface{}, error) {
+		return r.SuggestedRoutesV2(ctx, input)
+	}, func(result interface{}, taskType async.TaskType, err error) {
+		signal.SendWalletEvent(signal.SuggestedRoutes, result)
```
handle `err`. We normally include an error code in the event's payload, for example
```
func (s *Service) FilterActivityAsync(requestID int32, addresses []common.Address, chainIDs []w_common.ChainID, filter Filter, offset int, limit int) {
	s.scheduler.Enqueue(requestID, filterTask, func(ctx context.Context) (interface{}, error) {
		allAddresses := s.areAllAddresses(addresses)
		activities, err := getActivityEntries(ctx, s.getDeps(), addresses, allAddresses, chainIDs, filter, offset, limit)
		return activities, err
	}, func(result interface{}, taskType async.TaskType, err error) {
		res := FilterResponse{
			ErrorCode: ErrorCodeFailed,
		}

		if errors.Is(err, context.Canceled) || errors.Is(err, async.ErrTaskOverwritten) {
			res.ErrorCode = ErrorCodeTaskCanceled
		} else if err == nil {
			activities := result.([]Entry)
			res.Activities = activities
			res.Offset = offset
			res.HasMore = len(activities) == limit
			res.ErrorCode = ErrorCodeSuccess
		}

		sendResponseEvent(s.eventFeed, &requestID, EventActivityFilteringDone, res, err)

		s.getActivityDetailsAsync(requestID, res.Activities)
	})
}
```

dlipicar replied:
Also, I don't think it's required now for this case, but in case you feel it's convenient to add it: with these Async API endpoints I think it's generally useful to have some `uuid` that's set by the caller that's included as part of the response (so you can know on the client side which Event matches which request).

saledjenic replied:
> We normally include an error code in the event's payload

Yes, I noticed that while testing the changes, thanks for the heads up.

> I think it's generally useful to have some uuid

In this context it's not explicitly needed, but maybe a good time to add it for future needs and possibly give an option to clients for more flexible implementation.

ilmotta commented:
> @ilmotta changes done here should not affect the mobile app since it doesn't respond to `sing-transactions` signal (you will start using it when you implement keycard tx signing).

@qfrank, since you will be touching transaction signing via keycard, this is probably of interest to you.

ilmotta commented on services/wallet/router/router_v2.go:
```diff
@@ -417,26 +418,41 @@ func validateInputData(input *RouteInputParams) error {
 }
 
+func (r *Router) SuggestedRoutesV2Async(input *RouteInputParams) {
+	r.scheduler.Enqueue(async.TaskType{
```
@saledjenic, I'm just curious more than anything, so no suggestions for change.

1. I'm assuming we are going in the async path because calculating the routes can take a while, correct? Does the queue preemptively kill tasks that take too long? Is it "guaranteed" that getting suggested routes will finish eventually and within a reasonable amount of time?
2. I don't know if it would be useful, but in case getting suggested routes takes too long, clients may find useful to receive a token/identifier from this endpoint that they can later use at their own discretion to cancel enqueued tasks (for example to dispatch again `SuggestedRoutesV2Async` if they detect connection issues or based on some other state outside status-go).

saledjenic replied:
> I'm assuming we are going in the async path because calculating the routes can take a while, correct?

Correct.

> Does the queue preemptively kill tasks that take too long?

You can see that the set policy is `ReplacementPolicyCancelOld`, which means the current task will be canceled and replaced with a new one, you can see the implementation details here https://github.com/status-im/status-go/blob/develop/services/wallet/async/scheduler.go#L67-L133

> Is it "guaranteed" that getting suggested routes will finish eventually and within a reasonable amount of time?

The router algorithm does many external calls, and for each of them we have a timeout set, so if something takes long will be canceled after the timeout.

> I don't know if it would be useful, but in case getting suggested routes takes too long, clients may find it useful to receive a token/identifier from this endpoint that they can later use at their own discretion to cancel enqueued tasks

The UI is designed in a way that the user cannot ask for the route and then close that popup and switch to some other view. But you're probably referring to if the user wants to close that popup we should explicitly close the running task, correct? That's not how we're doing it now, we're closing the popup, task remains till the end of its execution, we just don't respond to the received signal in that case (that's how it is/was). 

But thinking, since we're doing some changes now, maybe it's a good time to introduce that and will be on the client how will use it, whether to leave task running (ignoring the signal) or explicitly cancel it.

> for example to dispatch again SuggestedRoutesV2Async if they detect connection issues

If there are connection issues the route can't be found, cause the first external call after disconnection will return an error.

ilmotta replied:
Thank you @saledjenic for the insightful answers.

> maybe it's a good time to introduce that and will be on the client how will use it, whether to leave task running (ignoring the signal) or explicitly cancel it

At least on mobile, when a signal arrives and we ignore it, it still takes resources because it has to go through a queue processed by re-frame (a lib managing all sorts of events in the app) which competes for attention from other UI events in this single-threaded environment, among other details. Nothing too important, but whenever a signal is not published it helps a bit with performance.

I know my comment about making this (potentially) long-running signal cancellable  is leaning on the side of over-engineering, so please ignore it if it feels that way.

saledjenic replied:
It's already added, and there is now the `StopSuggestedRoutesV2AsyncCalcualtion` endpoint you can call from the mobile app when the user is about to close the sending page.

alaibe reviewed (approved):
uuid sounds good to me!

IvanBelyakoff reviewed (approved):
LGTM!

saledjenic commented:
I added `uuid` for async calls and also added an endpoint to stop async tasks, it can be used by the client when the sending modal is about to be closed.

dlipicar commented on services/wallet/router/router_v2.go:
```diff
@@ -33,6 +34,7 @@ var (
 )
 
 type RouteInputParams struct {
+	Uuid                 int64                   `json:"uuid"`
```
we normally use `string` for uuid in other parts of the app (https://en.wikipedia.org/wiki/Universally_unique_identifier)

saledjenic committed:
feat_: the router returns route there is a balance on even that's not the cheapest route

If there are multiple routes across multiple networks, but the user doesn't have a positive balance
on the network which the router initially suggested as the best (cheapest) route, then we are not returning
an error saying there are not enough balance, but instead try to suggest the route on the network where
the user has a positive balance even that's not the cheapest route (it should be the second cheapest route,
but if there are not enough balance on it we proceed with the third cheapest route and so on...).

saledjenic commented:
Please check the third commit here [this one](https://github.com/status-im/status-go/pull/5426/commits/b01af6107e019b8b8fea41cd56c0930d8986da1b), it's an important one, which is fixing the issue described here https://github.com/status-im/status-desktop/pull/15344#issuecomment-2196438083

It does this:
If there are multiple routes across multiple networks, but the user doesn't have a positive balance
on the network which the router initially suggested as the best (cheapest) route, then we are not returning
an error saying there is not enough balance, but instead try to suggest the route on the network where
the user has a positive balance even if that's not the cheapest route (it should be the second cheapest route,
but if there is not enough balance on it we proceed with the third cheapest route and so on...).

dlipicar committed:
fix_: fix swap in routerV2
